File: hhp227study/week_14/Baekjoon2606.java
/**
바이러스
https://www.acmicpc.net/problem/2606
*/

import java.util.*;

public class Main {
    private static final List<String> dfsResults = new ArrayList<>();

    private static final List<String> bfsResults = new ArrayList<>();

    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);
        int n = scanner.nextInt(), m = scanner.nextInt();
        Map<Integer, List<Integer>> map = new HashMap<>() {
            {
                for (int i = 0; i < n; i++) {
                    put(i + 1, new ArrayList<>());
                }
                for (int j = 0; j < m; j++) {
                    int[] inputs = new int[] {scanner.nextInt(), scanner.nextInt()};

                    get(inputs[0]).add(inputs[1]);
                    get(inputs[1]).add(inputs[0]);
                }
            }
        };

        dfs(map, 1, new Stack<>(), new HashSet<>(List.of(1)));
        bfs(map, 1, new LinkedList<>(), new HashSet<>(List.of(1)));
        if (dfsResults.size() == bfsResults.size()) {
            System.out.println(dfsResults.size() - 1);
        }
        scanner.close();
    }

    private static void dfs(Map<Integer, List<Integer>> edgeMap, int v, Stack<Integer> stack, Set<Integer> visited) {
        dfsResults.add(String.valueOf(v));
        for (int next : edgeMap.get(v)) {
            if (visited.contains(next)) {
                continue;
            }
            stack.push(next);
            visited.add(next);
            if (!stack.isEmpty()) {
                dfs(edgeMap, next, stack, visited);
            }
            stack.pop();
        }
    }

    private static void bfs(Map<Integer, List<Integer>> edgeMap, int v, Queue<Integer> queue, Set<Integer> visited) {
        queue.offer(v);
        while (!queue.isEmpty()) {
            int outed = queue.poll();

            bfsResults.add(String.valueOf(outed));
            visited.add(outed);
            for (int next : edgeMap.get(outed)) {
                if (!visited.contains(next) && !queue.contains(next)) {
                    queue.offer(next);
                }
            }
        }
    }
}

File: hhp227study/hhp227study_2023/week_02/Baekjoon2606.kt
/**
바이러스
https://www.acmicpc.net/problem/2606
*/

fun main() {
    val n = readln().toInt()
    val m = readln().toInt()
    val map = (1..n).associateWith { mutableListOf<Int>() }.apply {
        repeat(m) {
            readln().split(" ").map(String::toInt).also { (a, b) ->
                getOrDefault(a, mutableListOf()).add(b)
                getOrDefault(b, mutableListOf()).add(a)
            }
        }
    }
    val visited = mutableSetOf(1)
    fun dfs(v: Int) {
        for (next in map.getOrDefault(v, mutableListOf())) {
            if (visited.contains(next)) continue
            visited.add(next)
            dfs(next)
        }
    }

    dfs(1)
    println(visited.size - 1)
}
